al_util.js

```diff
@@ -7,137 +7,146 @@
-    // Default values
-    // randomize: false
-    factor: 7,
-    minTimeout: 300,
-    retries: 2,
-    maxTimeout: 10000,
-    maxRetryTime: 180000 // Maximum time to spend retrying in milliseconds.
+  // Default values
+  // randomize: false
+  factor: 7,
+  minTimeout: 300,
+  retries: 2,
+  maxTimeout: 10000,
+  maxRetryTime: 180000 
 };
 
 /**
- * @function Default retry callback. 
- *  It doesn't retry 2XX, 3XX and 4XX.
+ * @default Refer to httpAgent and httpAgents axios
+ */
+let DEFAULT_HTTP_HTTPS_AGENT_CONFIG = {
+  keepAlive: true,
+  maxSockets: MAX_CONNS_PER_SERVICE,   // Maximum number of sockets to open
+};
+
+
+/**
+ * @function Default retry callback.
+ *  It doesn't retry 2XX, 3XX, and 4XX.
  *  Keeps retrying 5XX HTTP responses and any system level errors
  **/
-var defaultRetryCb = function(err){
-    if (err && 
-        (err.statusCode >= 500 ||
-        (err.error && err.error.errno))) {
-        return true;
-    } else {
-        return false;
-    }
+var defaultRetryCb = function (err) {
+  if (err && (err.status >= 500) || (err.response && err.response.status >= 500) || (err.errno)) {
+    return true;
+  } else {
+    return false;
+  }
 };
 
 /**
  * @class
  * Rest client.
  *
  * @constructor
- * @param {string} endpoint - hostname/address to sent HTTPS
- * requests to.
+ * @param {string} endpoint - hostname/address to sent HTTPS requests to.
  *
  */
 class RestServiceClient {
-    constructor(endpoint, retryOptions) {
-        'use strict';
-        this._host = endpoint;
-        this._url = 'https://' + endpoint;
-        this._pool = {
-            maxSockets: MAX_CONNS_PER_SERVICE
-        };
-        if (retryOptions) {
-            this._retryCb = retryOptions.retryCb ? retryOptions.retryCb : defaultRetryCb;
-            delete retryOptions.retryCb;
-            this._retryOptions = retryOptions ? retryOptions : DEFAULT_RETRY;
-        } else {
-            this._retryCb = defaultRetryCb;
-            this._retryOptions = DEFAULT_RETRY;
-        }
+  constructor(endpoint, retryOptions) {
+    'use strict';
+    this._host = endpoint;
+    this._url = 'https://' + endpoint;
+    if (retryOptions) {
+      this._retryCb = retryOptions.retryCb ? retryOptions.retryCb : defaultRetryCb;
+      delete retryOptions.retryCb;
+      this._retryOptions = retryOptions ? retryOptions : DEFAULT_RETRY;
+    } else {
+      this._retryCb = defaultRetryCb;
+      this._retryOptions = DEFAULT_RETRY;
     }
+  }
+
+  _initRequestOptions(method, path, extra) {
+    'use strict';
+    const defaultOptions = {
+      method: method,
+      url: this._url + path,
+      headers: {},
+      json:true,
+      httpAgent: new http.Agent(DEFAULT_HTTP_HTTPS_AGENT_CONFIG),
+      httpsAgent: new https.Agent(DEFAULT_HTTP_HTTPS_AGENT_CONFIG)
+    };
+    const options = Object.assign({}, defaultOptions, extra);
+    const defaultHeaders = {
+      'Accept': 'application/json'
+    };
+    options.data = options.body;
```

Comment:
In some request, we don't need body or data object . Or if some one directly set data object.
In both scenario it will set data to undefined. 
It would be good to have conditional set. May be like 
```
if(!options.data || options.body){
options.data = options.body
}
```